libcudacxx/include/cuda/std/detail/libcxx/include/__utility/pair.h

```diff
@@ -69,707 +70,704 @@
-#endif // defined(__cuda_std__) && !defined(__CUDACC_RTC__)
-
-    template<class _U1, class _U2, __enable_if_t<
-            _CheckArgs::template __enable_implicit<_U1, _U2>::value
-    >* = nullptr>
-    _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11
-    pair(pair<_U1, _U2>&& __p)
-        noexcept((is_nothrow_constructible<first_type, _U1&&>::value &&
-                    is_nothrow_constructible<second_type, _U2&&>::value))
-        : first(_CUDA_VSTD::forward<_U1>(__p.first)), second(_CUDA_VSTD::forward<_U2>(__p.second)) {}
-
-#if defined(__cuda_std__) && !defined(__CUDACC_RTC__)
-    template<class _U1, class _U2, __enable_if_t<
-            _CheckArgs::template __enable_implicit<_U1, _U2>::value
-    >* = nullptr>
-    _LIBCUDACXX_HOST _LIBCUDACXX_CONSTEXPR_AFTER_CXX11
-    pair(::std::pair<_U1, _U2>&& __p)
-        noexcept((is_nothrow_constructible<first_type, _U1&&>::value &&
-                    is_nothrow_constructible<second_type, _U2&&>::value))
-        : first(_CUDA_VSTD::forward<_U1>(__p.first)), second(_CUDA_VSTD::forward<_U2>(__p.second)) {}
-#endif // defined(__cuda_std__) && !defined(__CUDACC_RTC__)
-
-#if _CCCL_STD_VER > 2020
-    template<class _U1, class _U2, __enable_if_t<
-            _CheckArgs::template __is_pair_constructible<const _U1&&, const _U2&&>::value
-    >* = nullptr>
-    _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY constexpr
-    explicit(!_CheckArgs::template __is_implicit<const _U1&&, const _U2&&>::value)
-    pair(const pair<_U1, _U2>&& __p)
-        noexcept(is_nothrow_constructible<first_type, const _U1&&>::value &&
-                 is_nothrow_constructible<second_type, const _U2&&>::value)
-        : first(_CUDA_VSTD::move(__p.first)), second(_CUDA_VSTD::move(__p.second)) {}
-
-#if defined(__cuda_std__) && !defined(__CUDACC_RTC__)
-    template<class _U1, class _U2, __enable_if_t<
-            _CheckArgs::template __is_pair_constructible<const _U1&&, const _U2&&>::value
-    >* = nullptr>
-    _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_HOST constexpr
-    explicit(!_CheckArgs::template __is_implicit<const _U1&&, const _U2&&>::value)
-    pair(const ::std::pair<_U1, _U2>&& __p)
-        noexcept(is_nothrow_constructible<first_type, const _U1&&>::value &&
-                 is_nothrow_constructible<second_type, const _U2&&>::value)
-        : first(_CUDA_VSTD::move(__p.first)), second(_CUDA_VSTD::move(__p.second)) {}
-#endif // defined(__cuda_std__) && !defined(__CUDACC_RTC__)
-#endif // _CCCL_STD_VER > 2020
+struct __pair_constraints
+{
+  static constexpr bool __implicit_default_constructible =
+    __is_implicitly_default_constructible<_T1>::value && __is_implicitly_default_constructible<_T2>::value;
+
+  static constexpr bool __explicit_default_constructible =
+    !__implicit_default_constructible && _LIBCUDACXX_TRAIT(is_default_constructible, _T1)
+    && _LIBCUDACXX_TRAIT(is_default_constructible, _T2);
+
+  static constexpr bool __implicit_copy_constructible_from_T =
```

Comment:
So at first (just reading the traits) I thought that these 4 should have the `_from_T` suffix removed from their names, but seeing how they are used, I'd suggest something like `__implicit_constructible_from_element_copies` and so on; it might be longer, but since there's no `T` in scope here, even logically, I don't think the current names make much sense. (`_from_Ts` maybe, but I think that's still muddled in its meaning.)

Reply:
The thing I probably dislike the most about the current name is that it contains "copy_constructible". This has a meaning - this thing we are constructing is created using its copy constructor. But that's not what's happening here, hence my original suggestion.

Reply:
Oh yeah, we should just name that `__{implicit,explicit}_constructible_from_elements`